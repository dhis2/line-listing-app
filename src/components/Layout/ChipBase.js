import { DIMENSION_ID_ORGUNIT } from '@dhis2/analytics'
import i18n from '@dhis2/d2-i18n'
import cx from 'classnames'
import PropTypes from 'prop-types'
import React from 'react'
import {
    VALUE_TYPE_BOOLEAN,
    VALUE_TYPE_TRUE_ONLY,
    getConditions,
} from '../../modules/conditions.js'
import { DIMENSION_TYPE_PERIOD } from '../../modules/dimensionConstants.js'
import { DimensionIcon } from '../MainSidebar/DimensionItem/DimensionIcon.js'
import styles from './styles/Chip.module.css'

const VALUE_TYPE_TRUE_ONLY_NUM_OPTIONS = 2
const VALUE_TYPE_BOOLEAN_NUM_OPTIONS = 3

// Presentational component used by dnd - do not add redux or dnd functionality

export const ChipBase = ({ dimension, conditions, items, metadata }) => {
    const { id, name, dimensionType, optionSet, valueType, stageName } =
        dimension

    const getChipSuffix = () => {
        if (
            (id === DIMENSION_ID_ORGUNIT ||
                dimensionType === DIMENSION_TYPE_PERIOD) &&
            !items.length
        ) {
            return ''
        }

        const all = i18n.t('all')

        if (!conditions.condition && !conditions.legendSet && !items.length) {
            return `: ${all}`
        }

        const conds = getConditions({ conditions, metadata, dimension })

        if (
            (valueType === VALUE_TYPE_TRUE_ONLY &&
                conds.length === VALUE_TYPE_TRUE_ONLY_NUM_OPTIONS) ||
            (valueType === VALUE_TYPE_BOOLEAN &&
                conds.length === VALUE_TYPE_BOOLEAN_NUM_OPTIONS)
        ) {
            return `: ${all}`
        }

        const numberOfConditions = conds.length

        if (optionSet || items.length) {
            const selected = items.length || numberOfConditions
            const suffix = i18n.t('{{count}} selected', {
                count: selected,
                defaultValue: '{{count}} selected',
                defaultValue_plural: '{{count}} selected',
            })
            return `: ${suffix}`
        } else if (numberOfConditions) {
            const suffix = i18n.t('{{count}} conditions', {
                count: numberOfConditions,
                defaultValue: '{{count}} condition',
                defaultValue_plural: '{{count}} conditions',
            })
            return `: ${suffix}`
        }

        return ''
    }

    return (
        <div className={cx(styles.chipBase)}>
            <div className={styles.leftIcon}>
                <DimensionIcon dimensionType={dimensionType} />
            </div>
            <span className={styles.label}>
                <span className={styles.primary}>{name}</span>
                {stageName && (
                    <span className={styles.secondary}>{stageName}</span>
                )}
            </span>
            <span className={styles.suffix}>{getChipSuffix()}</span>
        </div>
    )
}

ChipBase.propTypes = {
    conditions: PropTypes.object,
    dimension: PropTypes.shape({
        dimensionType: PropTypes.string,
        id: PropTypes.string,
        name: PropTypes.string,
        optionSet: PropTypes.string,
        stageName: PropTypes.string,
        valueType: PropTypes.string,
    }),
    items: PropTypes.array,
    metadata: PropTypes.object,
}

ChipBase.defaultProps = {
    conditions: {},
    items: [],
    metadata: {},
}
