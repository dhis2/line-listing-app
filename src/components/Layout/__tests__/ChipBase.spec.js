import { render } from '@testing-library/react'
import React from 'react'
import { ChipBase } from '../ChipBase.js'

describe('ChipBase', () => {
    test('Period: none selected', () => {
        const dimension = {
            id: 'eventDate',
            name: 'Event date (e2e)',
            dimensionType: 'PERIOD',
        }

        const { container } = render(
            <ChipBase
                dimension={dimension}
                conditionsLength={0}
                itemsLength={0}
            />
        )
        expect(container).toMatchSnapshot()
    })
    test('Period: 1 selected', () => {
        const dimension = {
            id: 'eventDate',
            name: 'Event date (e2e)',
            dimensionType: 'PERIOD',
        }

        const { container } = render(
            <ChipBase
                dimension={dimension}
                conditionsLength={0}
                itemsLength={1}
            />
        )
        expect(container).toMatchSnapshot()
    })
    test('OU: none selected', () => {
        const dimension = {
            id: 'ou',
            name: 'Organsiation unit',
            dimensionType: 'ORGANISATION_UNIT',
        }

        const { container } = render(
            <ChipBase
                dimension={dimension}
                conditionsLength={0}
                itemsLength={0}
            />
        )
        expect(container).toMatchSnapshot()
    })
    test('OU: 2 selected', () => {
        const dimension = {
            id: 'ou',
            name: 'Organsiation unit',
            dimensionType: 'ORGANISATION_UNIT',
        }

        const { container } = render(
            <ChipBase
                dimension={dimension}
                conditionsLength={0}
                itemsLength={2}
            />
        )
        expect(container).toMatchSnapshot()
    })
    test('Data element: no conditions', () => {
        const dimension = {
            id: 'dataElementId',
            name: 'My data element',
            dimensionType: 'DATA_ELEMENT',
            valueType: 'NUMBER',
        }

        const { container } = render(
            <ChipBase
                dimension={dimension}
                conditionsLength={0}
                itemsLength={0}
            />
        )
        expect(container).toMatchSnapshot()
    })
    // TODO: figure out why plural translations are not being returned from i18n.t() in Jest tests
    // The snapshot should say "2 conditions"
    test('Data element: 2 conditions', () => {
        const dimension = {
            id: 'dataElementId',
            name: 'My data element',
            dimensionType: 'DATA_ELEMENT',
            valueType: 'NUMBER',
        }

        const { container } = render(
            <ChipBase
                dimension={dimension}
                conditionsLength={2}
                itemsLength={0}
            />
        )
        expect(container).toMatchSnapshot()
    })
    test('Data element: option set and 2 options chosen', () => {
        const dimension = {
            id: 'dataElementId',
            name: 'My data element',
            dimensionType: 'DATA_ELEMENT',
            valueType: 'NUMBER',
            optionSet: 'optionSetId',
        }

        const { container } = render(
            <ChipBase
                dimension={dimension}
                conditionsLength={2}
                itemsLength={0}
            />
        )
        expect(container).toMatchSnapshot()
    })
    test('Data element: in stage with option set and 2 options chosen', () => {
        const dimension = {
            id: 'stageId.dataElementId',
            name: 'My data element',
            dimensionType: 'DATA_ELEMENT',
            valueType: 'NUMBER',
            optionSet: 'optionSetId',
            stageName: 'Stage 1',
        }

        const { container } = render(
            <ChipBase
                dimension={dimension}
                conditionsLength={2}
                itemsLength={0}
            />
        )
        expect(container).toMatchSnapshot()
    })
    test('Data element: TRUE_ONLY no conditions', () => {
        const dimension = {
            id: 'dataElementId',
            name: 'My data element',
            dimensionType: 'DATA_ELEMENT',
            valueType: 'TRUE_ONLY',
        }

        const { container } = render(
            <ChipBase
                dimension={dimension}
                conditionsLength={0}
                itemsLength={0}
            />
        )
        expect(container).toMatchSnapshot()
    })
    test('Data element: TRUE_ONLY 1 condition', () => {
        const dimension = {
            id: 'dataElementId',
            name: 'My data element',
            dimensionType: 'DATA_ELEMENT',
            valueType: 'TRUE_ONLY',
        }

        const { container } = render(
            <ChipBase
                dimension={dimension}
                conditionsLength={1}
                itemsLength={0}
            />
        )
        expect(container).toMatchSnapshot()
    })
    test('Data element: TRUE_ONLY both selected', () => {
        const dimension = {
            id: 'dataElementId',
            name: 'My data element',
            dimensionType: 'DATA_ELEMENT',
            valueType: 'TRUE_ONLY',
        }

        const { container } = render(
            <ChipBase
                dimension={dimension}
                conditionsLength={2}
                itemsLength={0}
            />
        )
        expect(container).toMatchSnapshot()
    })
})
