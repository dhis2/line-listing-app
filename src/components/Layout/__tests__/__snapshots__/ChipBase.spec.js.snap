// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChipBase Data element: 2 conditions 1`] = `
<div>
  <div
    class="chipBase"
  >
    <div
      class="leftIcon"
    >
      <svg
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 1a1 1 0 011 1v11a1 1 0 01-1 1H4a1 1 0 01-1-1V2a1 1 0 011-1zm-8 9v3h8v-3zm2 1v1H5v-1zm0-4v1H5V7zm6-5H4v3h8zM6 3v1H5V3zM4 9h8V6H4z"
          fill="currentColor"
        />
      </svg>
    </div>
    <span
      class="label"
    >
      <span
        class="primary"
      >
        My data element
      </span>
    </span>
    <span
      class="items"
      data-test="chip-items"
    >
      2
    </span>
  </div>
</div>
`;

exports[`ChipBase Data element: TRUE_ONLY 1 condition 1`] = `
<div>
  <div
    class="chipBase"
  >
    <div
      class="leftIcon"
    >
      <svg
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 1a1 1 0 011 1v11a1 1 0 01-1 1H4a1 1 0 01-1-1V2a1 1 0 011-1zm-8 9v3h8v-3zm2 1v1H5v-1zm0-4v1H5V7zm6-5H4v3h8zM6 3v1H5V3zM4 9h8V6H4z"
          fill="currentColor"
        />
      </svg>
    </div>
    <span
      class="label"
    >
      <span
        class="primary"
      >
        My data element
      </span>
    </span>
    <span
      class="items"
      data-test="chip-items"
    >
      1
    </span>
  </div>
</div>
`;

exports[`ChipBase Data element: TRUE_ONLY both selected 1`] = `
<div>
  <div
    class="chipBase"
  >
    <div
      class="leftIcon"
    >
      <svg
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 1a1 1 0 011 1v11a1 1 0 01-1 1H4a1 1 0 01-1-1V2a1 1 0 011-1zm-8 9v3h8v-3zm2 1v1H5v-1zm0-4v1H5V7zm6-5H4v3h8zM6 3v1H5V3zM4 9h8V6H4z"
          fill="currentColor"
        />
      </svg>
    </div>
    <span
      class="label"
    >
      <span
        class="primary"
      >
        My data element
      </span>
    </span>
    <span
      class="items"
      data-test="chip-items"
    >
      all
    </span>
  </div>
</div>
`;

exports[`ChipBase Data element: TRUE_ONLY no conditions 1`] = `
<div>
  <div
    class="chipBase"
  >
    <div
      class="leftIcon"
    >
      <svg
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 1a1 1 0 011 1v11a1 1 0 01-1 1H4a1 1 0 01-1-1V2a1 1 0 011-1zm-8 9v3h8v-3zm2 1v1H5v-1zm0-4v1H5V7zm6-5H4v3h8zM6 3v1H5V3zM4 9h8V6H4z"
          fill="currentColor"
        />
      </svg>
    </div>
    <span
      class="label"
    >
      <span
        class="primary"
      >
        My data element
      </span>
    </span>
    <span
      class="items"
      data-test="chip-items"
    >
      all
    </span>
  </div>
</div>
`;

exports[`ChipBase Data element: in stage with option set and 2 options chosen 1`] = `
<div>
  <div
    class="chipBase"
  >
    <div
      class="leftIcon"
    >
      <svg
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 1a1 1 0 011 1v11a1 1 0 01-1 1H4a1 1 0 01-1-1V2a1 1 0 011-1zm-8 9v3h8v-3zm2 1v1H5v-1zm0-4v1H5V7zm6-5H4v3h8zM6 3v1H5V3zM4 9h8V6H4z"
          fill="currentColor"
        />
      </svg>
    </div>
    <span
      class="label"
    >
      <span
        class="primary"
      >
        My data element
      </span>
    </span>
    <span
      class="items"
      data-test="chip-items"
    >
      2
    </span>
  </div>
</div>
`;

exports[`ChipBase Data element: no conditions 1`] = `
<div>
  <div
    class="chipBase"
  >
    <div
      class="leftIcon"
    >
      <svg
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 1a1 1 0 011 1v11a1 1 0 01-1 1H4a1 1 0 01-1-1V2a1 1 0 011-1zm-8 9v3h8v-3zm2 1v1H5v-1zm0-4v1H5V7zm6-5H4v3h8zM6 3v1H5V3zM4 9h8V6H4z"
          fill="currentColor"
        />
      </svg>
    </div>
    <span
      class="label"
    >
      <span
        class="primary"
      >
        My data element
      </span>
    </span>
    <span
      class="items"
      data-test="chip-items"
    >
      all
    </span>
  </div>
</div>
`;

exports[`ChipBase Data element: option set and 2 options chosen 1`] = `
<div>
  <div
    class="chipBase"
  >
    <div
      class="leftIcon"
    >
      <svg
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 1a1 1 0 011 1v11a1 1 0 01-1 1H4a1 1 0 01-1-1V2a1 1 0 011-1zm-8 9v3h8v-3zm2 1v1H5v-1zm0-4v1H5V7zm6-5H4v3h8zM6 3v1H5V3zM4 9h8V6H4z"
          fill="currentColor"
        />
      </svg>
    </div>
    <span
      class="label"
    >
      <span
        class="primary"
      >
        My data element
      </span>
    </span>
    <span
      class="items"
      data-test="chip-items"
    >
      2
    </span>
  </div>
</div>
`;

exports[`ChipBase OU: 2 selected 1`] = `
<div>
  <div
    class="chipBase"
  >
    <div
      class="leftIcon"
    >
      <svg
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6 1v4H4v2h6V6h4v4h-4V8H4v4h6v-1h4v4h-4v-2H3.5a.5.5 0 01-.492-.41L3 12.5V5H2V1zm7 11h-2v2h2zm0-5h-2v2h2zM5 2H3v2h2z"
          fill="currentColor"
        />
      </svg>
    </div>
    <span
      class="label"
    >
      <span
        class="primary"
      >
        Organsiation unit
      </span>
    </span>
    <span
      class="items"
      data-test="chip-items"
    >
      2
    </span>
  </div>
</div>
`;

exports[`ChipBase OU: none selected 1`] = `
<div>
  <div
    class="chipBase"
  >
    <div
      class="leftIcon"
    >
      <svg
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6 1v4H4v2h6V6h4v4h-4V8H4v4h6v-1h4v4h-4v-2H3.5a.5.5 0 01-.492-.41L3 12.5V5H2V1zm7 11h-2v2h2zm0-5h-2v2h2zM5 2H3v2h2z"
          fill="currentColor"
        />
      </svg>
    </div>
    <span
      class="label"
    >
      <span
        class="primary"
      >
        Organsiation unit
      </span>
    </span>
  </div>
</div>
`;

exports[`ChipBase Period: 1 selected 1`] = `
<div>
  <div
    class="chipBase"
  >
    <div
      class="leftIcon"
    >
      <svg
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M11.5 2a.5.5 0 01.5.5V4h1a1 1 0 011 1v8a1 1 0 01-1 1H3a1 1 0 01-1-1V5a1 1 0 011-1h1V2.5a.5.5 0 011 0V4h6V2.5a.5.5 0 01.5-.5zM13 6H3v7h10zm-8 4a1 1 0 110 2 1 1 0 010-2zm3 0a1 1 0 110 2 1 1 0 010-2zM5 7a1 1 0 110 2 1 1 0 010-2zm3 0a1 1 0 110 2 1 1 0 010-2zm3 0a1 1 0 110 2 1 1 0 010-2z"
          fill="currentColor"
        />
      </svg>
    </div>
    <span
      class="label"
    >
      <span
        class="primary"
      >
        Event date (e2e)
      </span>
    </span>
    <span
      class="items"
      data-test="chip-items"
    >
      1
    </span>
  </div>
</div>
`;

exports[`ChipBase Period: none selected 1`] = `
<div>
  <div
    class="chipBase"
  >
    <div
      class="leftIcon"
    >
      <svg
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M11.5 2a.5.5 0 01.5.5V4h1a1 1 0 011 1v8a1 1 0 01-1 1H3a1 1 0 01-1-1V5a1 1 0 011-1h1V2.5a.5.5 0 011 0V4h6V2.5a.5.5 0 01.5-.5zM13 6H3v7h10zm-8 4a1 1 0 110 2 1 1 0 010-2zm3 0a1 1 0 110 2 1 1 0 010-2zM5 7a1 1 0 110 2 1 1 0 010-2zm3 0a1 1 0 110 2 1 1 0 010-2zm3 0a1 1 0 110 2 1 1 0 010-2z"
          fill="currentColor"
        />
      </svg>
    </div>
    <span
      class="label"
    >
      <span
        class="primary"
      >
        Event date (e2e)
      </span>
    </span>
  </div>
</div>
`;
